ha/setup/create_pacemaker_resources.py

```diff
@@ -0,0 +1,251 @@
+    SimpleCommand().run_cmd(cmd_s3auth)
+
+    if push:
+        cib_push(cib_xml)
+
+
+def haproxy(cib_xml, push=False):
+    """Create haproxy clone resource in pacemaker."""
+    cmd_haproxy = f"pcs -f {cib_xml} resource create haproxy systemd:haproxy clone op monitor interval=30"
+    SimpleCommand().run_cmd(cmd_haproxy)
+
+    if push:
+        cib_push(cib_xml)
+
+
+def sspl(cib_xml, push=False):
+    """Create sspl clone resource in pacemaker."""
+    # Using sspl-ll service file according to the content of SSPL repo
+    cmd_sspl = f"pcs -f {cib_xml} resource create sspl-ll systemd:sspl-ll clone op monitor interval=30"
+    SimpleCommand().run_cmd(cmd_sspl)
+
+    if push:
+        cib_push(cib_xml)
+
+
+def io_stack(cib_xml, push=False):
+    """Wrapper to create IO stack related resources."""
+    # TODO: Clarify where shall be sspl in this sequence
+    resources = [ motr_hax, s3auth, haproxy, s3servers, s3bp]
+    for rcs in resources:
+        rcs(cib_xml, push)
+
+def kibana_vip(cib_xml, vip, iface, cidr=24, push=False):
+    """Creates Virtual IP resource for kibana."""
+    cmd = f"pcs -f {cib_xml} resource create kibana-vip ocf:heartbeat:IPaddr2 \
+ip={vip} cidr_netmask={cidr} nic={iface} iflabel=v1 \
+op start   interval=0s timeout=60s \
+op monitor interval=5s timeout=20s \
+op stop    interval=0s timeout=60s"
+    SimpleCommand().run_cmd(cmd)
+
+    if push:
+        cib_push(cib_xml)
+
+def csm_kibana(cib_xml, push=False):
+    """Creates mandatory resources for mgmt stack."""
+    kibana = f"pcs -f {cib_xml} resource create kibana systemd:kibana op monitor interval=30s"
+    agent = f"pcs -f {cib_xml} resource create csm-agent systemd:csm_agent op monitor interval=30s"
+    web = f"pcs -f {cib_xml} resource create csm-web systemd:csm_web op monitor interval=30s"
+
+    for c in (kibana, agent, web):
+        SimpleCommand().run_cmd(c)
+
+    if push:
+        cib_push(cib_xml)
+
+def uds(cib_xml, push=False):
+    """Create uds resource and constraints."""
+    cmd_uds = f"pcs -f {cib_xml} resource create uds systemd:uds op monitor interval=30s"
+    SimpleCommand().run_cmd(cmd_uds)
+    constraints = [
+            f"pcs -f {cib_xml} constraint colocation add uds with csm-agent score=INFINITY",
+            # According to EOS-9258, there is a bug which requires UDS to be started after csm_agent
+            f"pcs -f {cib_xml} constraint order csm-agent then uds"
+            ]
+    for c in constraints:
+        SimpleCommand().run_cmd(c)
+    if push:
+        cib_push(cib_xml)
+
+
+def mgmt_stack(cib_xml, kibana_vip_cfg, push=False):
+    """Wrapper to create Mgmt stack related resources.
+
+    It also creates and defines csm-kibana group to support colocation and
+    ordering requirements."""
+    csm_kibana(cib_xml)
+
+    uds(cib_xml)
+
+    cmd = f"pcs -f {cib_xml} resource group add csm-kibana kibana csm-agent csm-web"
```

Comment:
If we define these in a group then all the services will be started on same node?
If yes, then remove the grouping else rename csm-kibana to management(mgmt).

Reply:
All these resources are co-located together according to HA requirement document. As a consequence - yes, they always will be started together.
LR1 code also introduced ordering between uds, csm-agent and csm-web (AFAIR). So, since we have ordering and colocation - group is merely a syntactic sugar here.

Are colocation requirements in HA document still relevant?

Reply:
Yes. Those are relavant. Sorry for the confusion.